File: webapp/src/types/application.ts
// Mirrors backend/app/models/application.py::ApplicationStatus.
// Kept as a plain string-literal union (not a TS enum) so it serializes
// identically to the JSON the API sends/expects — no extra mapping layer.
export type ApplicationStatus =
  | 'saved'
  | 'applied'
  | 'phone_screen'
  | 'interviewing'
  | 'offer'
  | 'rejected'
  | 'withdrawn'
  | 'accepted'

// Order matters here: this is the left-to-right column order the Kanban
// board (Phase 2, later step) will use, and a sensible top-to-bottom order
// for a <select> filter too.
export const APPLICATION_STATUSES: readonly ApplicationStatus[] = [
  'saved',
  'applied',
  'phone_screen',
  'interviewing',
  'offer',
  'accepted',
  'rejected',
  'withdrawn',
]

export const APPLICATION_STATUS_LABELS: Record<ApplicationStatus, string> = {
  saved: 'Saved',
  applied: 'Applied',
  phone_screen: 'Phone Screen',
  interviewing: 'Interviewing',
  offer: 'Offer',
  accepted: 'Accepted',
  rejected: 'Rejected',
  withdrawn: 'Withdrawn',
}

// Mirrors backend/app/models/application.py::SalaryCurrency. A curated set
// of major job-market currencies, not the full ISO 4217 list — extend here
// (and in the backend enum + migration) if a currency outside this set
// shows up on a job.
export type SalaryCurrency =
  | 'USD'
  | 'EUR'
  | 'GBP'
  | 'CAD'
  | 'AUD'
  | 'NZD'
  | 'CHF'
  | 'SEK'
  | 'NOK'
  | 'DKK'
  | 'ISK'
  | 'PLN'
  | 'CZK'
  | 'HUF'
  | 'RON'
  | 'UAH'
  | 'RUB'
  | 'TRY'
  | 'ILS'
  | 'AED'
  | 'SAR'
  | 'EGP'
  | 'NGN'
  | 'KES'
  | 'ZAR'
  | 'INR'
  | 'PKR'
  | 'BDT'
  | 'CNY'
  | 'JPY'
  | 'KRW'
  | 'TWD'
  | 'HKD'
  | 'SGD'
  | 'MYR'
  | 'THB'
  | 'VND'
  | 'IDR'
  | 'PHP'
  | 'BRL'
  | 'MXN'
  | 'ARS'
  | 'CLP'
  | 'COP'

export const SALARY_CURRENCIES: readonly SalaryCurrency[] = [
  'USD',
  'EUR',
  'GBP',
  'CAD',
  'AUD',
  'NZD',
  'CHF',
  'SEK',
  'NOK',
  'DKK',
  'ISK',
  'PLN',
  'CZK',
  'HUF',
  'RON',
  'UAH',
  'RUB',
  'TRY',
  'ILS',
  'AED',
  'SAR',
  'EGP',
  'NGN',
  'KES',
  'ZAR',
  'INR',
  'PKR',
  'BDT',
  'CNY',
  'JPY',
  'KRW',
  'TWD',
  'HKD',
  'SGD',
  'MYR',
  'THB',
  'VND',
  'IDR',
  'PHP',
  'BRL',
  'MXN',
  'ARS',
  'CLP',
  'COP',
]

export const SALARY_CURRENCY_LABELS: Record<SalaryCurrency, string> = {
  USD: 'US Dollar',
  EUR: 'Euro',
  GBP: 'British Pound',
  CAD: 'Canadian Dollar',
  AUD: 'Australian Dollar',
  NZD: 'New Zealand Dollar',
  CHF: 'Swiss Franc',
  SEK: 'Swedish Krona',
  NOK: 'Norwegian Krone',
  DKK: 'Danish Krone',
  ISK: 'Icelandic Krona',
  PLN: 'Polish Zloty',
  CZK: 'Czech Koruna',
  HUF: 'Hungarian Forint',
  RON: 'Romanian Leu',
  UAH: 'Ukrainian Hryvnia',
  RUB: 'Russian Ruble',
  TRY: 'Turkish Lira',
  ILS: 'Israeli Shekel',
  AED: 'UAE Dirham',
  SAR: 'Saudi Riyal',
  EGP: 'Egyptian Pound',
  NGN: 'Nigerian Naira',
  KES: 'Kenyan Shilling',
  ZAR: 'South African Rand',
  INR: 'Indian Rupee',
  PKR: 'Pakistani Rupee',
  BDT: 'Bangladeshi Taka',
  CNY: 'Chinese Yuan',
  JPY: 'Japanese Yen',
  KRW: 'South Korean Won',
  TWD: 'Taiwan Dollar',
  HKD: 'Hong Kong Dollar',
  SGD: 'Singapore Dollar',
  MYR: 'Malaysian Ringgit',
  THB: 'Thai Baht',
  VND: 'Vietnamese Dong',
  IDR: 'Indonesian Rupiah',
  PHP: 'Philippine Peso',
  BRL: 'Brazilian Real',
  MXN: 'Mexican Peso',
  ARS: 'Argentine Peso',
  CLP: 'Chilean Peso',
  COP: 'Colombian Peso',
}

// Symbol shown alongside salary_min/salary_max in the application form —
// not necessarily the symbol used in-country, but the one job seekers
// scanning a salary figure will recognize fastest.
export const SALARY_CURRENCY_SYMBOLS: Record<SalaryCurrency, string> = {
  USD: '$',
  EUR: '€',
  GBP: '£',
  CAD: 'C$',
  AUD: 'A$',
  NZD: 'NZ$',
  CHF: 'CHF',
  SEK: 'kr',
  NOK: 'kr',
  DKK: 'kr',
  ISK: 'kr',
  PLN: 'zł',
  CZK: 'Kč',
  HUF: 'Ft',
  RON: 'lei',
  UAH: '₴',
  RUB: '₽',
  TRY: '₺',
  ILS: '₪',
  AED: 'AED',
  SAR: 'SAR',
  EGP: 'E£',
  NGN: '₦',
  KES: 'KSh',
  ZAR: 'R',
  INR: '₹',
  PKR: '₨',
  BDT: '৳',
  CNY: 'CN¥',
  JPY: '¥',
  KRW: '₩',
  TWD: 'NT$',
  HKD: 'HK$',
  SGD: 'S$',
  MYR: 'RM',
  THB: '฿',
  VND: '₫',
  IDR: 'Rp',
  PHP: '₱',
  BRL: 'R$',
  MXN: 'MX$',
  ARS: 'AR$',
  CLP: 'CL$',
  COP: 'CO$',
}

// Mirrors ApplicationRead (backend/app/schemas/application.py).
export interface Application {
  id: string
  user_id: string
  company: string
  position: string
  /** Optional user-chosen label to tell apart applications to the same
   * company/position (a re-apply, two postings with the same title). */
  application_name: string | null
  location: string | null
  status: ApplicationStatus
  salary_min: number | null
  salary_max: number | null
  salary_currency: SalaryCurrency
  /** ISO date string, e.g. "2026-07-16" (backend `date`, not `datetime`). */
  applied_date: string | null
  job_url: string | null
  notes: string | null
  created_at: string
  updated_at: string
}

// Mirrors ApplicationListResponse.
export interface ApplicationListResponse {
  items: Application[]
  total: number
  page: number
  page_size: number
}

// Query params accepted by GET /applications (backend uses `status` as the
// query alias, `status_filter` is just the Python-side param name).
//
// `status` is `ApplicationStatus | null | undefined`, not just optional:
// - `undefined` (key omitted) => "don't change this filter"
// - `null`                    => "explicitly clear this filter"
// This distinction matters for partial updates like `fetchApplications({ page: 2 })`,
// which should keep whatever status filter is already active rather than
// wiping it.
export interface ApplicationListParams {
  status?: ApplicationStatus | null
  search?: string
  page?: number
  page_size?: number
}

// Mirrors ApplicationCreate, which is just ApplicationBase — every field
// the backend accepts on create. `status` is optional since the backend
// defaults it to `saved`.
export interface ApplicationCreatePayload {
  company: string
  position: string
  application_name?: string | null
  location?: string | null
  status?: ApplicationStatus
  salary_min?: number | null
  salary_max?: number | null
  salary_currency?: SalaryCurrency
  applied_date?: string | null
  job_url?: string | null
  notes?: string | null
}

// Mirrors ApplicationUpdate: identical fields, all optional. The backend
// uses `exclude_unset=True`, so only keys actually present in the request
// body are touched — omit a field here rather than sending `undefined`/
// `null` for "don't change this".
export type ApplicationUpdatePayload = Partial<ApplicationCreatePayload>

File: mobile/lib/features/applications/domain/application.dart
/// Mirrors backend/app/models/application.py::ApplicationStatus.
///
/// Order matches APPLICATION_STATUSES in webapp/src/types/application.ts —
/// used for filter-list order and, later, any Kanban-style grouping.
enum ApplicationStatus {
  saved,
  applied,
  phoneScreen,
  interviewing,
  offer,
  accepted,
  rejected,
  withdrawn;

  /// The exact string the backend sends/expects. Not `name`, since Dart
  /// enum member names are camelCase (`phoneScreen`) but the API speaks
  /// snake_case (`phone_screen`).
  String get apiValue => switch (this) {
        ApplicationStatus.saved => 'saved',
        ApplicationStatus.applied => 'applied',
        ApplicationStatus.phoneScreen => 'phone_screen',
        ApplicationStatus.interviewing => 'interviewing',
        ApplicationStatus.offer => 'offer',
        ApplicationStatus.accepted => 'accepted',
        ApplicationStatus.rejected => 'rejected',
        ApplicationStatus.withdrawn => 'withdrawn',
      };

  /// Mirrors APPLICATION_STATUS_LABELS in webapp/src/types/application.ts.
  String get label => switch (this) {
        ApplicationStatus.saved => 'Saved',
        ApplicationStatus.applied => 'Applied',
        ApplicationStatus.phoneScreen => 'Phone Screen',
        ApplicationStatus.interviewing => 'Interviewing',
        ApplicationStatus.offer => 'Offer',
        ApplicationStatus.accepted => 'Accepted',
        ApplicationStatus.rejected => 'Rejected',
        ApplicationStatus.withdrawn => 'Withdrawn',
      };

  static ApplicationStatus fromApiValue(String value) {
    return ApplicationStatus.values.firstWhere(
      (status) => status.apiValue == value,
      orElse: () => throw FormatException('Unknown application status: $value'),
    );
  }
}

/// Mirrors backend/app/models/application.py::SalaryCurrency and
/// webapp/src/types/application.ts::SalaryCurrency — a curated set of
/// major job-market currencies, not the full ISO 4217 list. Extend here
/// (and in the backend enum + migration, and the webapp mirror) if a
/// currency outside this set shows up on a job.
enum SalaryCurrency {
  usd,
  eur,
  gbp,
  cad,
  aud,
  nzd,
  chf,
  sek,
  nok,
  dkk,
  isk,
  pln,
  czk,
  huf,
  ron,
  uah,
  rub,
  // `try` is a reserved Dart keyword, so the ISO code TRY can't be the
  // member name directly.
  tryLira,
  ils,
  aed,
  sar,
  egp,
  ngn,
  kes,
  zar,
  inr,
  pkr,
  bdt,
  cny,
  jpy,
  krw,
  twd,
  hkd,
  sgd,
  myr,
  thb,
  vnd,
  idr,
  php,
  brl,
  mxn,
  ars,
  clp,
  cop;

  /// The exact string the backend sends/expects.
  String get apiValue => switch (this) {
        SalaryCurrency.usd => 'USD',
        SalaryCurrency.eur => 'EUR',
        SalaryCurrency.gbp => 'GBP',
        SalaryCurrency.cad => 'CAD',
        SalaryCurrency.aud => 'AUD',
        SalaryCurrency.nzd => 'NZD',
        SalaryCurrency.chf => 'CHF',
        SalaryCurrency.sek => 'SEK',
        SalaryCurrency.nok => 'NOK',
        SalaryCurrency.dkk => 'DKK',
        SalaryCurrency.isk => 'ISK',
        SalaryCurrency.pln => 'PLN',
        SalaryCurrency.czk => 'CZK',
        SalaryCurrency.huf => 'HUF',
        SalaryCurrency.ron => 'RON',
        SalaryCurrency.uah => 'UAH',
        SalaryCurrency.rub => 'RUB',
        SalaryCurrency.tryLira => 'TRY',
        SalaryCurrency.ils => 'ILS',
        SalaryCurrency.aed => 'AED',
        SalaryCurrency.sar => 'SAR',
        SalaryCurrency.egp => 'EGP',
        SalaryCurrency.ngn => 'NGN',
        SalaryCurrency.kes => 'KES',
        SalaryCurrency.zar => 'ZAR',
        SalaryCurrency.inr => 'INR',
        SalaryCurrency.pkr => 'PKR',
        SalaryCurrency.bdt => 'BDT',
        SalaryCurrency.cny => 'CNY',
        SalaryCurrency.jpy => 'JPY',
        SalaryCurrency.krw => 'KRW',
        SalaryCurrency.twd => 'TWD',
        SalaryCurrency.hkd => 'HKD',
        SalaryCurrency.sgd => 'SGD',
        SalaryCurrency.myr => 'MYR',
        SalaryCurrency.thb => 'THB',
        SalaryCurrency.vnd => 'VND',
        SalaryCurrency.idr => 'IDR',
        SalaryCurrency.php => 'PHP',
        SalaryCurrency.brl => 'BRL',
        SalaryCurrency.mxn => 'MXN',
        SalaryCurrency.ars => 'ARS',
        SalaryCurrency.clp => 'CLP',
        SalaryCurrency.cop => 'COP',
      };

  /// Mirrors SALARY_CURRENCY_LABELS in webapp/src/types/application.ts.
  String get label => switch (this) {
        SalaryCurrency.usd => 'US Dollar',
        SalaryCurrency.eur => 'Euro',
        SalaryCurrency.gbp => 'British Pound',
        SalaryCurrency.cad => 'Canadian Dollar',
        SalaryCurrency.aud => 'Australian Dollar',
        SalaryCurrency.nzd => 'New Zealand Dollar',
        SalaryCurrency.chf => 'Swiss Franc',
        SalaryCurrency.sek => 'Swedish Krona',
        SalaryCurrency.nok => 'Norwegian Krone',
        SalaryCurrency.dkk => 'Danish Krone',
        SalaryCurrency.isk => 'Icelandic Krona',
        SalaryCurrency.pln => 'Polish Zloty',
        SalaryCurrency.czk => 'Czech Koruna',
        SalaryCurrency.huf => 'Hungarian Forint',
        SalaryCurrency.ron => 'Romanian Leu',
        SalaryCurrency.uah => 'Ukrainian Hryvnia',
        SalaryCurrency.rub => 'Russian Ruble',
        SalaryCurrency.tryLira => 'Turkish Lira',
        SalaryCurrency.ils => 'Israeli Shekel',
        SalaryCurrency.aed => 'UAE Dirham',
        SalaryCurrency.sar => 'Saudi Riyal',
        SalaryCurrency.egp => 'Egyptian Pound',
        SalaryCurrency.ngn => 'Nigerian Naira',
        SalaryCurrency.kes => 'Kenyan Shilling',
        SalaryCurrency.zar => 'South African Rand',
        SalaryCurrency.inr => 'Indian Rupee',
        SalaryCurrency.pkr => 'Pakistani Rupee',
        SalaryCurrency.bdt => 'Bangladeshi Taka',
        SalaryCurrency.cny => 'Chinese Yuan',
        SalaryCurrency.jpy => 'Japanese Yen',
        SalaryCurrency.krw => 'South Korean Won',
        SalaryCurrency.twd => 'Taiwan Dollar',
        SalaryCurrency.hkd => 'Hong Kong Dollar',
        SalaryCurrency.sgd => 'Singapore Dollar',
        SalaryCurrency.myr => 'Malaysian Ringgit',
        SalaryCurrency.thb => 'Thai Baht',
        SalaryCurrency.vnd => 'Vietnamese Dong',
        SalaryCurrency.idr => 'Indonesian Rupiah',
        SalaryCurrency.php => 'Philippine Peso',
        SalaryCurrency.brl => 'Brazilian Real',
        SalaryCurrency.mxn => 'Mexican Peso',
        SalaryCurrency.ars => 'Argentine Peso',
        SalaryCurrency.clp => 'Chilean Peso',
        SalaryCurrency.cop => 'Colombian Peso',
      };

  /// Mirrors SALARY_CURRENCY_SYMBOLS in webapp/src/types/application.ts —
  /// not necessarily the symbol used in-country, but the one job seekers
  /// scanning a salary figure will recognize fastest.
  String get symbol => switch (this) {
        SalaryCurrency.usd => '\$',
        SalaryCurrency.eur => '€',
        SalaryCurrency.gbp => '£',
        SalaryCurrency.cad => 'C\$',
        SalaryCurrency.aud => 'A\$',
        SalaryCurrency.nzd => 'NZ\$',
        SalaryCurrency.chf => 'CHF',
        SalaryCurrency.sek => 'kr',
        SalaryCurrency.nok => 'kr',
        SalaryCurrency.dkk => 'kr',
        SalaryCurrency.isk => 'kr',
        SalaryCurrency.pln => 'zł',
        SalaryCurrency.czk => 'Kč',
        SalaryCurrency.huf => 'Ft',
        SalaryCurrency.ron => 'lei',
        SalaryCurrency.uah => '₴',
        SalaryCurrency.rub => '₽',
        SalaryCurrency.tryLira => '₺',
        SalaryCurrency.ils => '₪',
        SalaryCurrency.aed => 'AED',
        SalaryCurrency.sar => 'SAR',
        SalaryCurrency.egp => 'E£',
        SalaryCurrency.ngn => '₦',
        SalaryCurrency.kes => 'KSh',
        SalaryCurrency.zar => 'R',
        SalaryCurrency.inr => '₹',
        SalaryCurrency.pkr => '₨',
        SalaryCurrency.bdt => '৳',
        SalaryCurrency.cny => 'CN¥',
        SalaryCurrency.jpy => '¥',
        SalaryCurrency.krw => '₩',
        SalaryCurrency.twd => 'NT\$',
        SalaryCurrency.hkd => 'HK\$',
        SalaryCurrency.sgd => 'S\$',
        SalaryCurrency.myr => 'RM',
        SalaryCurrency.thb => '฿',
        SalaryCurrency.vnd => '₫',
        SalaryCurrency.idr => 'Rp',
        SalaryCurrency.php => '₱',
        SalaryCurrency.brl => 'R\$',
        SalaryCurrency.mxn => 'MX\$',
        SalaryCurrency.ars => 'AR\$',
        SalaryCurrency.clp => 'CL\$',
        SalaryCurrency.cop => 'CO\$',
      };

  static SalaryCurrency fromApiValue(String value) {
    return SalaryCurrency.values.firstWhere(
      (currency) => currency.apiValue == value,
      orElse: () => throw FormatException('Unknown salary currency: $value'),
    );
  }
}

/// Mirrors ApplicationRead (backend/app/schemas/application.py) and
/// webapp/src/types/application.ts::Application.
class Application {
  const Application({
    required this.id,
    required this.userId,
    required this.company,
    required this.position,
    required this.applicationName,
    required this.location,
    required this.status,
    required this.salaryMin,
    required this.salaryMax,
    required this.salaryCurrency,
    required this.appliedDate,
    required this.jobUrl,
    required this.notes,
    required this.createdAt,
    required this.updatedAt,
  });

  final String id;
  final String userId;
  final String company;
  final String position;

  /// Optional, user-chosen label distinguishing applications that share
  /// the same company/position (e.g. a re-apply after rejection) —
  /// mirrors backend/app/models/application.py's application_name.
  final String? applicationName;
  final String? location;
  final ApplicationStatus status;
  final int? salaryMin;
  final int? salaryMax;
  final SalaryCurrency salaryCurrency;

  /// Backend sends a plain `date` (e.g. "2026-07-16"), not a `datetime` —
  /// parsed here but should only ever be used for its date components.
  final DateTime? appliedDate;
  final String? jobUrl;
  final String? notes;
  final DateTime createdAt;
  final DateTime updatedAt;

  factory Application.fromJson(Map<String, dynamic> json) {
    return Application(
      id: json['id'] as String,
      userId: json['user_id'] as String,
      company: json['company'] as String,
      position: json['position'] as String,
      applicationName: json['application_name'] as String?,
      location: json['location'] as String?,
      status: ApplicationStatus.fromApiValue(json['status'] as String),
      salaryMin: json['salary_min'] as int?,
      salaryMax: json['salary_max'] as int?,
      salaryCurrency:
          SalaryCurrency.fromApiValue(json['salary_currency'] as String),
      appliedDate: json['applied_date'] == null
          ? null
          : DateTime.parse(json['applied_date'] as String),
      jobUrl: json['job_url'] as String?,
      notes: json['notes'] as String?,
      createdAt: DateTime.parse(json['created_at'] as String),
      updatedAt: DateTime.parse(json['updated_at'] as String),
    );
  }
}

/// Mirrors ApplicationListResponse (backend/app/schemas/application.py).
class ApplicationListResponse {
  const ApplicationListResponse({
    required this.items,
    required this.total,
    required this.page,
    required this.pageSize,
  });

  final List<Application> items;
  final int total;
  final int page;
  final int pageSize;

  factory ApplicationListResponse.fromJson(Map<String, dynamic> json) {
    return ApplicationListResponse(
      items: (json['items'] as List<dynamic>)
          .map((item) => Application.fromJson(item as Map<String, dynamic>))
          .toList(),
      total: json['total'] as int,
      page: json['page'] as int,
      pageSize: json['page_size'] as int,
    );
  }
}
